### PR TITLE
Refactor build_op_downstream_map for standalone executor

### PR DESCRIPTION
**Pull request by From00**
### PR types
Others

### PR changes
Others

### Describe

1. dependency_utils.h/cc -> dependency_builder.h/cc
2. 将新执行器构建依赖代码从interpretercore_utils.h/cc统一移动到dependency_builder.h/cc
3. 将表达依赖关系的op_downstream_map数据类型从std::map<int, std::list<int>>修改为std::map<int, std::set<int>>，避免重边
4. AddDownstreamOp中加入实时更新op_happens_before_的逻辑，加强剪枝判断

### CODE REVIEW

File: paddle/fluid/framework/new_executor/interpreter/dependency_builder.h [hunk 1]
```diff
@@ -37,17 +23,44 @@ namespace paddle {
...
-    const std::vector<std::vector<bool>>* op_happens_before = nullptr);
+// DependencyBuilder provides some dependency adding function to handle the
+// dependency that cannot be explicitly expresed by a Program. It is a
+// compromise of the incomplete expression ability of the Program. Do not add
+// too many functions here at will, that will bring great burden to the
+// Interpretercore.
+
+class DependencyBuilder {
+ public:
+  explicit DependencyBuilder() : instructions_(nullptr) {}
```
Comment (zhiqiu) [inline paddle/fluid/framework/new_executor/interpreter/dependency_builder.h]: NIT, explicit is not needed here.

File: paddle/fluid/framework/new_executor/interpreter/dependency_builder.h [hunk 2]
```diff
@@ -37,17 +23,44 @@ namespace paddle {
...
+
+  const std::map<int, std::set<int>>& Build(
```
Comment (zhiqiu) [inline paddle/fluid/framework/new_executor/interpreter/dependency_builder.h:38]: Better add comments on the function, especially the  `return`, or, rename it to specify that it returns `op_downstream_map_`

File: paddle/fluid/framework/new_executor/interpreter/dependency_builder.h [hunk 3]
```diff
@@ -37,17 +23,44 @@ namespace paddle {
...
+
+  void ShrinkDownstreamMap();
+
+  bool is_build_{false};
+  const std::vector<Instruction>* instructions_;  // not_own
+  size_t op_num_;
+
+  // op_happens_before_[i][j] == true means op[i] happens before op[j]
+  std::vector<std::vector<bool>> op_happens_before_;
+  std::map<int, std::set<int>> op_downstream_map_;
```
Comment (zhiqiu) [inline paddle/fluid/framework/new_executor/interpreter/dependency_builder.h:68]: Explain the meaning of `op_downstream_map_[i] = {a, b, c}` as above

File: paddle/fluid/framework/new_executor/interpreter/dependency_builder.h (hunk 1 above)
Comment (From00) [reply paddle/fluid/framework/new_executor/interpreter/dependency_builder.h]: Done

File: paddle/fluid/framework/new_executor/interpreter/dependency_builder.h (hunk 2 above)
Comment (From00) [reply paddle/fluid/framework/new_executor/interpreter/dependency_builder.h:38]: Done

File: paddle/fluid/framework/new_executor/interpreter/dependency_builder.h (hunk 3 above)
Comment (From00) [reply paddle/fluid/framework/new_executor/interpreter/dependency_builder.h:68]: Done

**Review by zhiqiu — APPROVED** [review]
LGTM

**Review by phlrain — APPROVED** [review]

**Pull request merged**